sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/_evaluators/_document_retrieval/_document_retrieval.py

```diff
@@ -0,0 +1,342 @@
+
+        ranks = list(range(1, self.k + 1))
+        xdcg_n = sum(
+            starmap(
+                calculate_xdcg_numerator, zip(result_docs_groundtruth_labels, ranks)
+            )
+        )
+        xdcg_d = sum(map(calculate_xdcg_denominator, ranks))
+
+        return xdcg_n / float(xdcg_d)
+
+    def _compute_fidelity(
+        self,
+        result_docs_groundtruth_labels: List[int],
+        ideal_docs_groundtruth_labels: List[int],
+    ) -> float:
+        """
+        Fidelity calculated over all documents retrieved from a search query.
+        Fidelity measures how objectively good are all of the documents retrieved compared with all known good documents in the underlying data store.
+        """
+
+        def calculate_weighted_sum_by_rating(labels: List[int]) -> float:
+            s = self.groundtruth_label_min + 1
+
+            # get a count of each label
+            label_counts = {str(i): 0 for i in range(s, self.groundtruth_label_max + 1)}
+
+            for label in labels:
+                if label >= s:
+                    label_counts[str(label)] += 1
+
+            sorted_label_counts = [
+                x[1] for x in sorted(label_counts.items(), key=lambda x: x[0])
+            ]
+
+            # calculate weights
+            weights = [
+                (math.pow(2, i + 1) - 1)
+                for i in range(s, self.groundtruth_label_max + 1)
+            ]
+
+            # return weighted sum
+            return sum(starmap(operator.mul, zip(sorted_label_counts, weights)))
+
+        weighted_sum_by_rating_results = calculate_weighted_sum_by_rating(
+            result_docs_groundtruth_labels
+        )
+        weighted_sum_by_rating_index = calculate_weighted_sum_by_rating(
+            ideal_docs_groundtruth_labels
+        )
+
+        if weighted_sum_by_rating_index == 0:
+            return math.nan
+
+        return weighted_sum_by_rating_results / float(weighted_sum_by_rating_index)
+
+    def _get_binary_result(self, **metrics) -> Dict[str, float]:
+        result = {}
+
+        for metric_name, metric_value in metrics.items():
+            if metric_name in self._threshold_metrics.keys():
+                result[f"{metric_name}_result"] = (
+                    metric_value >= self._threshold_metrics[metric_name]
+                )
+                result[f"{metric_name}_threshold"] = self._threshold_metrics[
+                    metric_name
+                ]
+                result[f"{metric_name}_lower_is_better"] = False
+
+            elif metric_name in self._threshold_holes.keys():
+                result[f"{metric_name}_result"] = (
+                    metric_value <= self._threshold_holes[metric_name]
+                )
+                result[f"{metric_name}_threshold"] = self._threshold_holes[metric_name]
+                result[f"{metric_name}_lower_is_better"] = True
+
+            else:
+                raise ValueError(f"No threshold set for metric '{metric_name}'")
```

Comment:
Can we add info on how to set threshold if it is missing ?

Reply:
I actually think this case should be considered a bug for us to fix if it ever occurs.  We should at the very least be setting a default value in the class, so if this error occurs it means that we missed one.  I'll add a test case to be sure that it doesn't happen.